COMMIT-MESSAGES.md

```diff
@@ -0,0 +1,102 @@
+- If the summary line is too long, intermediate segments may be omitted or abbreviated if they contribute no meaningful specificity. The resulting path should still intuitively lead the reader to the change, while freeing more space for the summary message.
+
+#### Summary phrasing
+
+- The summary message must appear after the final path segment's colon. It must describe the nature of the change, preferably beginning with a verb.
+
+#### Formatting and readability
+
+- Do *not* wrap path segments (or any code) in quotes or backticks.
+- Write summaries as plain text--GitHub's rendering is a distraction, not a target.
+- Do *not* employ visual markup tricks, which waste space and can be misleading.
+- Only include special characters in paths and messages when they are syntactically meaningful in the language.
+
+If you find it difficult to summarize a commit using this format, it's a strong sign that the change should be split into smaller, more focused commits--perhaps even re-ordered.
+
+# Why not *Conventional Commits*?
+
+- The verb-prefixes don't add meaningful value to the commit log and don't help with reviewability.
+- Changelogs should start from a list of commits since the last tagged release and be curated by a human to communicate relevant information to consumers.
+- The prefixes aren't used as part of a readable phrase, even though they could be incorporated more naturally.
+- *Conventional Commits* don't encourage small, focused commits and may lead to sprawling changes being squashed or created as a single commit.
+- The format prioritizes the developer's problem statement over the actual implementation detail or architectural change.
+- Problems should be explained in pull-request descriptions, not shoehorned into commit summaries.
+- Commit message bodies can explain why each change builds on the previous to achieve the intended result.
+- *Conventional Commit* summaries rarely indicate *where* a change was made--likely because they're too broad.
+- Ad-hoc scope identifiers in parentheses (`feat(foo):`) can conflict with past or future commits. Even when specific enough to be useful, they still introduce visual noise and consume precious space in the summary line.
+
+## Examples
+
+### Change to a single file at root:
+
+```
+readme: reword the description
+```
+```
+license: change from Apache-2.0 to MIT
+```
+
+### Change in a specific file or Rust module
+
+The file changed is in `src/foo/bar.rs` and the fully qualified module path (in the public API) is `foo::bar::baz`, where each of `foo`, `bar`, `baz` are public modules.
+
+```
+foo: bar: baz: rename fn do_thing to do_many_things
+```
+
+```
+foo: bar: baz: do_many_things: replace loop with try_fold
+```
+
+### Common change among multiple semantic paths
+
+```
+flake: devShells+packages: fix rust-overlay toolchain usage
```

Comment:
AFAIU `devShells` and `packages` are "children" of `flake`. What would happen when two or more files have been changed in the root of the repo and cannot be divided into separate commits without compromising that commit's ability to build?

What I currently do is mention the "main" change in the commit summary and the rest of the changes in the body. This is bad for visibility, but I haven't found anything better.

I guess one could write `root: a+b: change x to y`

Reply:
Cool, makes sense.

I do use `treewide` sometimes, never `root`. I used it for the sake of the example.